File: backend/app/wechat/ilink_client.py
# -*- coding: utf-8 -*-
"""微信 iLink HTTP + JSON 客户端。"""
from __future__ import annotations

import base64
import secrets
from dataclasses import dataclass
from typing import Any

import httpx

from .models import ILinkCredentials


class ILinkError(Exception):
    """iLink 请求或协议错误。"""


@dataclass(frozen=True)
class ILinkConfig:
    base_url: str = 'https://ilinkai.weixin.qq.com'
    timeout_seconds: float = 40.0


class ILinkClient:
    def __init__(
        self,
        config: ILinkConfig | None = None,
        credentials: ILinkCredentials | None = None,
        http_client: httpx.AsyncClient | None = None,
    ):
        self.config = config or ILinkConfig()
        self.credentials = credentials
        self._http_client = http_client

    def set_credentials(self, credentials: ILinkCredentials | None):
        self.credentials = credentials
        if credentials and credentials.base_url:
            self.config = ILinkConfig(
                base_url=credentials.base_url.rstrip('/'),
                timeout_seconds=self.config.timeout_seconds,
            )

    async def get_bot_qrcode(self, bot_type: int = 3) -> dict[str, Any]:
        return await self._request('GET', 'get_bot_qrcode', params={'bot_type': bot_type}, auth=False)

    async def get_qrcode_status(self, qrcode: str) -> dict[str, Any]:
        return await self._request('GET', 'get_qrcode_status', params={'qrcode': qrcode}, auth=False)

    async def get_updates(self, get_updates_buf: str = '') -> dict[str, Any]:
        return await self._request('POST', 'getupdates', {'get_updates_buf': get_updates_buf})

    async def send_message(self, to_user_id: str, context_token: str, text: str) -> dict[str, Any]:
        return await self._request('POST', 'sendmessage', {
            'msg': {
                'to_user_id': to_user_id,
                'context_token': context_token,
                'item_list': [{'type': 1, 'text_item': {'text': text}}],
            }
        })

    async def get_config(self, ilink_user_id: str, context_token: str = '') -> dict[str, Any]:
        return await self._request('POST', 'getconfig', {
            'ilink_user_id': ilink_user_id,
            'context_token': context_token,
        })

    async def send_typing(self, ilink_user_id: str, typing_ticket: str, status: int) -> dict[str, Any]:
        return await self._request('POST', 'sendtyping', {
            'ilink_user_id': ilink_user_id,
            'typing_ticket': typing_ticket,
            'status': status,
        })

    async def close(self):
        if self._http_client is not None:
            await self._http_client.aclose()

    async def _request(
        self,
        method: str,
        endpoint: str,
        body: dict[str, Any] | None = None,
        *,
        params: dict[str, Any] | None = None,
        auth: bool = True,
    ) -> dict[str, Any]:
        headers = {'Content-Type': 'application/json'}
        if auth:
            if not self.credentials or not self.credentials.bot_token:
                raise ILinkError('微信尚未完成扫码授权')
            headers.update({
                'AuthorizationType': 'ilink_bot_token',
                'Authorization': f'Bearer {self.credentials.bot_token}',
                'X-WECHAT-UIN': _random_uin(),
            })
        url = f'{self.config.base_url.rstrip("/")}/ilink/bot/{endpoint}'
        client = self._http_client
        owns_client = client is None
        if owns_client:
            client = httpx.AsyncClient(timeout=self.config.timeout_seconds)
        try:
            response = await client.request(method, url, headers=headers, params=params, json=body)
            if response.status_code >= 400:
                raise ILinkError(f'iLink 返回 HTTP {response.status_code}: {response.text[:300]}')
            try:
                data = response.json()
            except ValueError as exc:
                raise ILinkError('iLink 返回了无效 JSON') from exc
            if isinstance(data, dict) and data.get('ret') not in (None, 0):
                raise ILinkError(f"iLink 错误 {data.get('errcode', data.get('ret'))}: {data.get('errmsg', '')}")
            return data
        except httpx.HTTPError as exc:
            raise ILinkError(f'iLink 网络请求失败：{exc}') from exc
        finally:
            if owns_client:
                await client.aclose()


def _random_uin() -> str:
    value = str(secrets.randbelow(2**32))
    return base64.b64encode(value.encode('ascii')).decode('ascii')

File: server/src/wechat/ilinkClient.ts
import { randomInt, randomUUID } from 'node:crypto';
import type { ILinkCredentials } from './models.js';

export const ILINK_APP_ID = 'bot';
export const ILINK_APP_CLIENT_VERSION = String((2 << 16) | (4 << 8) | 6);
export const ILINK_CHANNEL_VERSION = '2.4.6';
export const ILINK_BOT_AGENT = 'MeimeiWorkbench/1.0.0';

const DEFAULT_ILINK_BASE_URL = 'https://ilinkai.weixin.qq.com';

export class ILinkError extends Error {}

export class ILinkSessionExpiredError extends ILinkError {}

export interface ILinkConfig {
  base_url?: string;
  timeout_seconds?: number;
}

export interface HttpResponse {
  readonly status: number;
  text(): Promise<string>;
}

export type HttpFetch = (
  url: string,
  init: { method: string; headers: Record<string, string>; body?: string; signal?: AbortSignal },
) => Promise<HttpResponse>;

const defaultFetch: HttpFetch = (url, init) => fetch(url, {
  method: init.method,
  headers: init.headers,
  body: init.body,
  signal: init.signal,
});

export class ILinkClient {
  config: ILinkConfig;
  credentials: ILinkCredentials | null;
  httpClient: HttpFetch | null;
  private controller: AbortController | null = null;

  constructor(config: ILinkConfig | null = null, credentials: ILinkCredentials | null = null, httpClient: HttpFetch | null = null) {
    this.config = config ?? {};
    this.credentials = credentials;
    this.httpClient = httpClient;
  }

  setCredentials(credentials: ILinkCredentials | null): void {
    this.credentials = credentials;
    if (credentials && credentials.base_url) {
      this.config = {
        base_url: credentials.base_url.replace(/\/+$/, ''),
        timeout_seconds: this.config.timeout_seconds,
      };
    }
  }

  async getBotQrcode(botType = 3): Promise<Record<string, unknown>> {
    return this.request('GET', 'get_bot_qrcode', null, { params: { bot_type: botType }, auth: false });
  }

  async getQrcodeStatus(qrcode: string): Promise<Record<string, unknown>> {
    return this.request('GET', 'get_qrcode_status', null, { params: { qrcode }, auth: false });
  }

  async getUpdates(getUpdatesBuf = ''): Promise<Record<string, unknown>> {
    return this.request('POST', 'getupdates', { get_updates_buf: getUpdatesBuf });
  }

  async sendMessage(toUserId: string, contextToken: string, text: string): Promise<Record<string, unknown>> {
    return this.request('POST', 'sendmessage', {
      msg: {
        from_user_id: '',
        to_user_id: toUserId,
        client_id: `meimei-workbench-${randomUUID().replace(/-/g, '')}`,
        message_type: 2,
        message_state: 2,
        context_token: contextToken,
        item_list: [{ type: 1, text_item: { text } }],
      },
    });
  }

  async getConfig(ilinkUserId: string, contextToken = ''): Promise<Record<string, unknown>> {
    return this.request('POST', 'getconfig', {
      ilink_user_id: ilinkUserId,
      context_token: contextToken,
    });
  }

  async sendTyping(ilinkUserId: string, typingTicket: string, status: number): Promise<Record<string, unknown>> {
    return this.request('POST', 'sendtyping', {
      ilink_user_id: ilinkUserId,
      typing_ticket: typingTicket,
      status,
    });
  }

  abort(): void {
    if (this.controller) {
      this.controller.abort();
      this.controller = null;
    }
  }

  async close(): Promise<void> {
    this.abort();
  }

  private async request(
    method: string,
    endpoint: string,
    body: Record<string, unknown> | null = null,
    options: { params?: Record<string, unknown> | null; auth?: boolean } = {},
  ): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'iLink-App-Id': ILINK_APP_ID,
      'iLink-App-ClientVersion': ILINK_APP_CLIENT_VERSION,
    };
    let requestBody: Record<string, unknown> | null = body;
    if (options.auth !== false) {
      if (!this.credentials || !this.credentials.bot_token) {
        throw new ILinkError('微信尚未完成扫码授权');
      }
      headers['AuthorizationType'] = 'ilink_bot_token';
      headers['Authorization'] = `Bearer ${this.credentials.bot_token}`;
      headers['X-WECHAT-UIN'] = randomUin();
      if (body !== null) {
        requestBody = {
          ...body,
          base_info: {
            channel_version: ILINK_CHANNEL_VERSION,
            bot_agent: ILINK_BOT_AGENT,
          },
        };
      }
    }
    const baseUrl = (this.config.base_url ?? DEFAULT_ILINK_BASE_URL).replace(/\/+$/, '');
    const query = options.params
      ? '?' + new URLSearchParams(
        Object.entries(options.params).map(([key, value]) => [key, String(value)] as [string, string]),
      ).toString()
      : '';
    const url = `${baseUrl}/ilink/bot/${endpoint}${query}`;
    const http = this.httpClient ?? defaultFetch;
    const controller = new AbortController();
    this.controller = controller;
    try {
      let response: HttpResponse;
      try {
        response = await http(url, {
          method,
          headers,
          body: requestBody !== null ? JSON.stringify(requestBody) : undefined,
          signal: AbortSignal.any([
            controller.signal,
            AbortSignal.timeout(Math.max(1000, Math.ceil((this.config.timeout_seconds ?? 40) * 1000))),
          ]),
        });
      } catch (error) {
        throw new ILinkError(`iLink 网络请求失败：${(error as Error).message}`);
      }
      if (response.status >= 400) {
        const text = (await response.text()).slice(0, 300);
        throw new ILinkError(`iLink 返回 HTTP ${response.status}: ${text}`);
      }
      const raw = await response.text();
      if (!raw.trim()) return {};
      let data: unknown;
      try {
        data = JSON.parse(raw);
      } catch {
        throw new ILinkError('iLink 返回了无效 JSON');
      }
      if (typeof data === 'object' && data !== null) {
        const record = data as Record<string, unknown>;
        const ret = record.ret;
        if (ret !== undefined && ret !== null && ret !== 0) {
          const code = record.errcode ?? ret;
          const message = `iLink 错误 ${code}: ${String(record.errmsg ?? '')}`;
          if (String(code) === '-14') {
            throw new ILinkSessionExpiredError('微信 iLink 会话已过期，请重新扫码登录');
          }
          throw new ILinkError(message);
        }
      }
      return (data ?? {}) as Record<string, unknown>;
    } finally {
      if (this.controller === controller) this.controller = null;
    }
  }
}

function randomUin(): string {
  const value = String(randomInt(2 ** 32));
  return Buffer.from(value, 'ascii').toString('base64');
}
